sidecar/src/service/session_info.rs

```diff
@@ -0,0 +1,249 @@
+        match runtimes.get(runtime_id) {
+            Some(runtime) => runtime.clone(),
+            None => {
+                let mut runtime = crate::interface::RuntimeInfo::default();
+                runtimes.insert(runtime_id.clone(), runtime.clone());
+                #[cfg(feature = "tracing")]
+                if enabled!(Level::INFO) {
+                    runtime.instance_id = InstanceId {
+                        session_id: self.session_id.clone(),
+                        runtime_id: runtime_id.clone(),
+                    };
+                    info!(
+                        "Registering runtime_id {} for session {}",
+                        runtime_id, self.session_id
+                    );
+                }
+                runtime
+            }
+        }
+    }
+
+    /// Shuts down all runtimes in the session.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_sidecar::service::SessionInfo;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let session_info = SessionInfo::default();
+    ///     session_info.shutdown().await;
+    /// }
+    /// ```
+    pub async fn shutdown(&self) {
+        let runtimes: Vec<crate::interface::RuntimeInfo> = self
+            .lock_runtimes()
+            .drain()
+            .map(|(_, instance)| instance)
+            .collect();
+
+        let runtimes_shutting_down: Vec<_> = runtimes
+            .into_iter()
+            .map(|rt| tokio::spawn(async move { rt.shutdown().await }))
+            .collect();
+
+        future::join_all(runtimes_shutting_down).await;
+    }
+    /// Shuts down all running instances in the session.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_sidecar::service::SessionInfo;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let session_info = SessionInfo::default();
+    ///     session_info.shutdown_running_instances().await;
+    /// }
+    /// ```
+    pub async fn shutdown_running_instances(&self) {
+        let runtimes: Vec<crate::interface::RuntimeInfo> = self
+            .lock_runtimes()
+            .drain()
+            .map(|(_, instance)| instance)
+            .collect();
+
+        let instances_shutting_down: Vec<_> = runtimes
+            .into_iter()
+            .map(|rt| tokio::spawn(async move { rt.shutdown().await }))
+            .collect();
+
+        future::join_all(instances_shutting_down).await;
+    }
+
+    /// Shuts down a specific runtime in the session.
+    ///
+    /// # Arguments
+    ///
+    /// * `runtime_id` - The ID of the runtime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_sidecar::service::SessionInfo;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let session_info = SessionInfo::default();
+    ///     session_info.shutdown_runtime(&"runtime1".to_string()).await;
+    /// }
+    /// ```
+    pub async fn shutdown_runtime(&self, runtime_id: &String) {
```

Comment:
Calling this function out as there is a minor change (in addition to being moved). This was the only public function that took `self` instead of `&self`. This meant that after shutting down a single runtime the `session_info` object was moved and you couldn't do anything else with it. I'm not that familiar with how `SessionInfo` should be used, but since it supports multiple runtimes it should be usable after a single one is shutdown.

Reply:
It didn't matter, because it implements Clone and all the relevant data within is Arc. But yes, using `&self` is correct.